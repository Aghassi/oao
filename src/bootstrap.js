// @flow

import { mainStory, chalk } from 'storyboard';
import kebabCase from 'kebab-case';
import semver from 'semver';
import { readAllSpecs, ROOT_PACKAGE } from './utils/readSpecs';
import removeInternalLinks from './utils/removeInternalLinks';
import writeSpecs from './utils/writeSpecs';
import { exec } from './utils/shell';

const PASS_THROUGH_OPTS = ['production', 'noLockfile', 'pureLockfile', 'frozenLockfile'];

type Options = {
  src: string,
  ignoreSrc?: string,
  link: ?string,
  production?: boolean,
};

const run = async (opts: Options) => {
  const { src, ignoreSrc, link: linkPattern } = opts;
  const production = opts.production || process.env.NODE_ENV === 'production';
  const allSpecs = await readAllSpecs(src, ignoreSrc);
  const pkgNames = Object.keys(allSpecs);
  const allRemovedDepsByPackage = {};
  const allRemovedDepsByPackageAndType = {};

  // Pass 1: register each package with yarn, and install external deps
  for (let i = 0; i < pkgNames.length; i++) {
    const pkgName = pkgNames[i];
    // if (pkgName === ROOT_PACKAGE) continue;
    const { displayName, pkgPath, specPath, specs: prevSpecs } = allSpecs[pkgName];
    mainStory.info(`${chalk.bold('PASS 1:')} processing ${chalk.cyan.bold(displayName)}...`);

    // Link
    if (pkgName !== ROOT_PACKAGE) {
      mainStory.info('  - Registering...');
      await exec('yarn link', { cwd: pkgPath, logLevel: 'trace', errorLogLevel: 'info' });
    }

    // Rewrite package.json without own/linked packages, install, and revert changes
    let fModified = false;
    try {
      const { nextSpecs, allRemovedPackages, removedPackagesByType } =
        removeInternalLinks(prevSpecs, pkgNames, linkPattern);
      allRemovedDepsByPackage[pkgName] = allRemovedPackages;
      allRemovedDepsByPackageAndType[pkgName] = removedPackagesByType;
      if (nextSpecs !== prevSpecs) {
        writeSpecs(specPath, nextSpecs);
        fModified = true;
      }
      mainStory.info('  - Installing external dependencies...');
      let cmd = 'yarn install';
      PASS_THROUGH_OPTS.forEach((key) => { if (opts[key]) cmd += ` --${kebabCase(key)}`; });
      await exec(cmd, { cwd: pkgPath, logLevel: 'trace' });
    } finally {
      if (prevSpecs != null && fModified) writeSpecs(specPath, prevSpecs);
    }
  }

  // Pass 2: link internal and user-specified deps
  for (let i = 0; i < pkgNames.length; i++) {
    const pkgName = pkgNames[i];
    const allRemovedPackages = allRemovedDepsByPackage[pkgName];
    const removedPackagesByType = allRemovedDepsByPackageAndType[pkgName];
    const packagesToLink = Object.keys(allRemovedPackages);
    const { displayName, pkgPath } = allSpecs[pkgName];
    mainStory.info(`${chalk.bold('PASS 2:')} installing internal deps for ` +
      `${chalk.cyan.bold(displayName)}...`);
    for (let k = 0; k < packagesToLink.length; k++) {
      const depName = packagesToLink[k];
      if (production && isPureDevDependency(removedPackagesByType, depName)) continue;
      mainStory.info(`  - Linking to ${chalk.cyan.bold(depName)}...`);
      const depVersionRange = allRemovedPackages[depName];
      const depSpecs = allSpecs[depName]; // might not exist, if it's a custom link
      const depActualVersion = depSpecs ? depSpecs.specs.version : null;
      if (depActualVersion && !semver.satisfies(depActualVersion, depVersionRange)) {
        mainStory.warn(`  - Warning: ${chalk.cyan.bold(`${depName}@${depActualVersion}`)} ` +
          `does not satisfy specified range: ${chalk.cyan.bold(depVersionRange)}`);
      }
      await exec(`yarn link ${depName}`, { cwd: pkgPath, logLevel: 'trace' });
    }
  }
};

const isPureDevDependency = (deps, depName) =>
  !((deps.dependencies && deps.dependencies[depName]) ||
  (deps.optionalDependencies && deps.optionalDependencies[depName]) ||
  (deps.peerDependencies && deps.peerDependencies[depName]));

export default run;
