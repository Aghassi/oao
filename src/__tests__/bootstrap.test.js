/* eslint-env jest */
/* eslint-disable global-require, import/no-dynamic-require */

import path from 'path';
import bootstrap from '../bootstrap';

jest.mock('../utils/shell');
jest.mock('../utils/writeSpecs');

const PACKAGE_NAMES_1 = ['oao', 'oao-b', 'oao-c', 'oao-d', 'oao-priv'];
const PACKAGE_NAMES_2 = ['oao', 'oao-b', 'oao-c'];

const readOriginalSpecs = (base, names) => {
  const originalSpecs = {};
  names.forEach((name) => {
    originalSpecs[name] = require(path.join(base, `${name}/package.json`));
  });
  return originalSpecs;
};

const spyFinalSpec = (spy, pkgName) => {
  let finalSpec;
  spy.mock.calls.forEach(([, specs]) => {
    if (specs.name === pkgName) finalSpec = specs;
  });
  return finalSpec;
};

describe('BOOTSTRAP command', () => {
  beforeEach(() => { jest.resetAllMocks(); });

  it('does not modify any package.json', async () => {
    const writeSpecs = require('../utils/writeSpecs').default;
    const base = path.join(process.cwd(), 'test/fixtures/packages');
    const originalSpecs = readOriginalSpecs(base, PACKAGE_NAMES_1);
    await bootstrap({ src: 'test/fixtures/packages/*' });
    // Process call arguments; we keep the last time a spec is stored
    // and compare it with the original one
    writeSpecs.mock.calls.forEach(([specPath, specs]) => {
      const { name } = specs;
      expect(specPath.split(path.sep)).toContain(name);
      expect(specPath.split(path.sep)).toContain('package.json');
    });
    PACKAGE_NAMES_1.forEach((name) => {
      expect(spyFinalSpec(writeSpecs, name)).toEqual(originalSpecs[name]);
    });
  });

  it('does not modify any package.json with custom links', async () => {
    const writeSpecs = require('../utils/writeSpecs').default;
    const base = path.join(process.cwd(), 'test/fixtures/packagesCustomLinks');
    const originalSpecs = readOriginalSpecs(base, PACKAGE_NAMES_2);
    await bootstrap({
      src: 'test/fixtures/packagesCustomLinks/*',
      link: 'ext-.*',
    });
    PACKAGE_NAMES_2.forEach((name) => {
      expect(spyFinalSpec(writeSpecs, name)).toEqual(originalSpecs[name]);
    });
  });

  it('executes the correct `yarn link`s and `yarn install`s', async () => {
    const helpers = require('../utils/shell');
    await bootstrap({ src: 'test/fixtures/packages/*' });
    expect(helpers.exec.mock.calls).toMatchSnapshot();
  });

  it('executes the correct `yarn link`s and `yarn install`s in production', async () => {
    const helpers = require('../utils/shell');
    await bootstrap({ src: 'test/fixtures/packages/*', production: true });
    expect(helpers.exec.mock.calls).toMatchSnapshot();
  });

  it('executes the correct `yarn link`s and `yarn install`s with custom links', async () => {
    const helpers = require('../utils/shell');
    await bootstrap({
      src: 'test/fixtures/packagesCustomLinks/*',
      link: 'ext-.*',
    });
    expect(helpers.exec.mock.calls).toMatchSnapshot();
  });

  it('supports scoped packages', async () => {
    const helpers = require('../utils/shell');
    await bootstrap({ src: 'test/fixtures/packagesScoped/*' });
    expect(helpers.exec.mock.calls).toMatchSnapshot();
  });
});
