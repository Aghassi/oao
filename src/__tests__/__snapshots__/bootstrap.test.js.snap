// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BOOTSTRAP command executes the correct \`yarn link\`s and \`yarn install\`s 1`] = `
Array [
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packages/oao",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install",
    Object {
      "cwd": "test/fixtures/packages/oao",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packages/oao-b",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install",
    Object {
      "cwd": "test/fixtures/packages/oao-b",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packages/oao-c",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install",
    Object {
      "cwd": "test/fixtures/packages/oao-c",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packages/oao-d",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install",
    Object {
      "cwd": "test/fixtures/packages/oao-d",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packages/oao-priv",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install",
    Object {
      "cwd": "test/fixtures/packages/oao-priv",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install",
    Object {
      "cwd": ".",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao",
    Object {
      "cwd": "test/fixtures/packages/oao-b",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao",
    Object {
      "cwd": "test/fixtures/packages/oao-c",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao-b",
    Object {
      "cwd": "test/fixtures/packages/oao-c",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao",
    Object {
      "cwd": "test/fixtures/packages/oao-d",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao-b",
    Object {
      "cwd": "test/fixtures/packages/oao-d",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao-c",
    Object {
      "cwd": "test/fixtures/packages/oao-d",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao",
    Object {
      "cwd": "test/fixtures/packages/oao-priv",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao-b",
    Object {
      "cwd": "test/fixtures/packages/oao-priv",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao-c",
    Object {
      "cwd": "test/fixtures/packages/oao-priv",
      "logLevel": "trace",
    },
  ],
]
`;

exports[`BOOTSTRAP command executes the correct \`yarn link\`s and \`yarn install\`s in production 1`] = `
Array [
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packages/oao",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install --production",
    Object {
      "cwd": "test/fixtures/packages/oao",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packages/oao-b",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install --production",
    Object {
      "cwd": "test/fixtures/packages/oao-b",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packages/oao-c",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install --production",
    Object {
      "cwd": "test/fixtures/packages/oao-c",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packages/oao-d",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install --production",
    Object {
      "cwd": "test/fixtures/packages/oao-d",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packages/oao-priv",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install --production",
    Object {
      "cwd": "test/fixtures/packages/oao-priv",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install --production",
    Object {
      "cwd": ".",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao",
    Object {
      "cwd": "test/fixtures/packages/oao-b",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao",
    Object {
      "cwd": "test/fixtures/packages/oao-c",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao-b",
    Object {
      "cwd": "test/fixtures/packages/oao-c",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao",
    Object {
      "cwd": "test/fixtures/packages/oao-d",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao",
    Object {
      "cwd": "test/fixtures/packages/oao-priv",
      "logLevel": "trace",
    },
  ],
]
`;

exports[`BOOTSTRAP command executes the correct \`yarn link\`s and \`yarn install\`s with --frozen-lockfile 1`] = `
Array [
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packages/oao",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install --frozen-lockfile",
    Object {
      "cwd": "test/fixtures/packages/oao",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packages/oao-b",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install --frozen-lockfile",
    Object {
      "cwd": "test/fixtures/packages/oao-b",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packages/oao-c",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install --frozen-lockfile",
    Object {
      "cwd": "test/fixtures/packages/oao-c",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packages/oao-d",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install --frozen-lockfile",
    Object {
      "cwd": "test/fixtures/packages/oao-d",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packages/oao-priv",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install --frozen-lockfile",
    Object {
      "cwd": "test/fixtures/packages/oao-priv",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install --frozen-lockfile",
    Object {
      "cwd": ".",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao",
    Object {
      "cwd": "test/fixtures/packages/oao-b",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao",
    Object {
      "cwd": "test/fixtures/packages/oao-c",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao-b",
    Object {
      "cwd": "test/fixtures/packages/oao-c",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao",
    Object {
      "cwd": "test/fixtures/packages/oao-d",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao-b",
    Object {
      "cwd": "test/fixtures/packages/oao-d",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao-c",
    Object {
      "cwd": "test/fixtures/packages/oao-d",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao",
    Object {
      "cwd": "test/fixtures/packages/oao-priv",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao-b",
    Object {
      "cwd": "test/fixtures/packages/oao-priv",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao-c",
    Object {
      "cwd": "test/fixtures/packages/oao-priv",
      "logLevel": "trace",
    },
  ],
]
`;

exports[`BOOTSTRAP command executes the correct \`yarn link\`s and \`yarn install\`s with --no-lockfile 1`] = `
Array [
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packages/oao",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install --no-lockfile",
    Object {
      "cwd": "test/fixtures/packages/oao",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packages/oao-b",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install --no-lockfile",
    Object {
      "cwd": "test/fixtures/packages/oao-b",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packages/oao-c",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install --no-lockfile",
    Object {
      "cwd": "test/fixtures/packages/oao-c",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packages/oao-d",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install --no-lockfile",
    Object {
      "cwd": "test/fixtures/packages/oao-d",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packages/oao-priv",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install --no-lockfile",
    Object {
      "cwd": "test/fixtures/packages/oao-priv",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install --no-lockfile",
    Object {
      "cwd": ".",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao",
    Object {
      "cwd": "test/fixtures/packages/oao-b",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao",
    Object {
      "cwd": "test/fixtures/packages/oao-c",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao-b",
    Object {
      "cwd": "test/fixtures/packages/oao-c",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao",
    Object {
      "cwd": "test/fixtures/packages/oao-d",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao-b",
    Object {
      "cwd": "test/fixtures/packages/oao-d",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao-c",
    Object {
      "cwd": "test/fixtures/packages/oao-d",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao",
    Object {
      "cwd": "test/fixtures/packages/oao-priv",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao-b",
    Object {
      "cwd": "test/fixtures/packages/oao-priv",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao-c",
    Object {
      "cwd": "test/fixtures/packages/oao-priv",
      "logLevel": "trace",
    },
  ],
]
`;

exports[`BOOTSTRAP command executes the correct \`yarn link\`s and \`yarn install\`s with --pure-lockfile 1`] = `
Array [
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packages/oao",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install --pure-lockfile",
    Object {
      "cwd": "test/fixtures/packages/oao",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packages/oao-b",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install --pure-lockfile",
    Object {
      "cwd": "test/fixtures/packages/oao-b",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packages/oao-c",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install --pure-lockfile",
    Object {
      "cwd": "test/fixtures/packages/oao-c",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packages/oao-d",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install --pure-lockfile",
    Object {
      "cwd": "test/fixtures/packages/oao-d",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packages/oao-priv",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install --pure-lockfile",
    Object {
      "cwd": "test/fixtures/packages/oao-priv",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install --pure-lockfile",
    Object {
      "cwd": ".",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao",
    Object {
      "cwd": "test/fixtures/packages/oao-b",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao",
    Object {
      "cwd": "test/fixtures/packages/oao-c",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao-b",
    Object {
      "cwd": "test/fixtures/packages/oao-c",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao",
    Object {
      "cwd": "test/fixtures/packages/oao-d",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao-b",
    Object {
      "cwd": "test/fixtures/packages/oao-d",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao-c",
    Object {
      "cwd": "test/fixtures/packages/oao-d",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao",
    Object {
      "cwd": "test/fixtures/packages/oao-priv",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao-b",
    Object {
      "cwd": "test/fixtures/packages/oao-priv",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao-c",
    Object {
      "cwd": "test/fixtures/packages/oao-priv",
      "logLevel": "trace",
    },
  ],
]
`;

exports[`BOOTSTRAP command executes the correct \`yarn link\`s and \`yarn install\`s with custom links 1`] = `
Array [
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packagesCustomLinks/oao",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install",
    Object {
      "cwd": "test/fixtures/packagesCustomLinks/oao",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packagesCustomLinks/oao-b",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install",
    Object {
      "cwd": "test/fixtures/packagesCustomLinks/oao-b",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packagesCustomLinks/oao-c",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install",
    Object {
      "cwd": "test/fixtures/packagesCustomLinks/oao-c",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install",
    Object {
      "cwd": ".",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao",
    Object {
      "cwd": "test/fixtures/packagesCustomLinks/oao-b",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link ext-one",
    Object {
      "cwd": "test/fixtures/packagesCustomLinks/oao-b",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao",
    Object {
      "cwd": "test/fixtures/packagesCustomLinks/oao-c",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link ext-one",
    Object {
      "cwd": "test/fixtures/packagesCustomLinks/oao-c",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link oao-b",
    Object {
      "cwd": "test/fixtures/packagesCustomLinks/oao-c",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link ext-two",
    Object {
      "cwd": "test/fixtures/packagesCustomLinks/oao-c",
      "logLevel": "trace",
    },
  ],
]
`;

exports[`BOOTSTRAP command supports scoped packages 1`] = `
Array [
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packagesScoped/example-package",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install",
    Object {
      "cwd": "test/fixtures/packagesScoped/example-package",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link",
    Object {
      "cwd": "test/fixtures/packagesScoped/example-package-b",
      "errorLogLevel": "info",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install",
    Object {
      "cwd": "test/fixtures/packagesScoped/example-package-b",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn install",
    Object {
      "cwd": ".",
      "logLevel": "trace",
    },
  ],
  Array [
    "yarn link @guigrpa/example-package",
    Object {
      "cwd": "test/fixtures/packagesScoped/example-package-b",
      "logLevel": "trace",
    },
  ],
]
`;
